fastapi/utils.py

```diff
@@ -258,3 +261,189 @@ def get_value_or_default(
         if not isinstance(item, DefaultPlaceholder):
             return item
     return first_item
+
+
+def _infer_type_from_ast(
+    node: ast.AST,
+    func_def: Union[ast.FunctionDef, ast.AsyncFunctionDef],
+    context_name: str,
+) -> Any:
+    if isinstance(node, ast.Constant):
+        return type(node.value)
+
+    if isinstance(node, ast.List):
+        if not node.elts:
+            return List[Any]
+
+        first_type = _infer_type_from_ast(node.elts[0], func_def, context_name + "Item")
+
+        for elt in node.elts[1:]:
+            current_type = _infer_type_from_ast(elt, func_def, context_name + "Item")
+            if current_type != first_type:
+                return List[Any]
+
+        if first_type is not Any:
+            return List[first_type]  # type: ignore
+        return List[Any]
+
+    if isinstance(node, ast.BinOp):
+        left_type = _infer_type_from_ast(node.left, func_def, context_name)
+        right_type = _infer_type_from_ast(node.right, func_def, context_name)
+        if left_type == right_type and left_type in (int, float, str):
+            return left_type
+        if {left_type, right_type} == {int, float}:
+            return float
+
+    if isinstance(node, ast.Compare):
+        return bool
+
+    if isinstance(node, ast.Dict):
+        fields = {}
+        for key, value in zip(node.keys, node.values):
+            if not isinstance(key, ast.Constant):
+                continue
+            field_name = key.value
+            field_type = _infer_type_from_ast(
+                value, func_def, context_name + "_" + str(field_name)
+            )
+            fields[field_name] = (field_type, ...)
+
+        if not fields:
+            return Dict[str, Any]
+
+        if PYDANTIC_V2:
+            from pydantic import create_model
+        else:
+            from pydantic import create_model
+
+        return create_model(f"Model_{context_name}", **fields)  # type: ignore[call-overload]
+
+    if isinstance(node, ast.Name):
+        arg_name = node.id
+        for arg in func_def.args.args:
+            if arg.arg == arg_name and arg.annotation:
+                if isinstance(arg.annotation, ast.Name):
+                    if arg.annotation.id == "int":
```

Comment:
This case (and quite a few other cases) makes use of the [Arrow Anti Pattern](https://wiki.c2.com/?ArrowAntiPattern), which renders the code quite a bit less readable! In this case, it'd be better to replace these nested conditionals with guard clauses. For example, at `if isinstance(node, ast.Name)`:

```python
if not isinstance(node, ast.Name):
    return Any
```